### PR TITLE
harden(executables): expand allowExecutables gate to mcp and canvas; store approvals user-local

### PR DESCRIPTION
**Pull request by sergio-sisternes-epam**
# harden(executables): expand allowExecutables gate to mcp and canvas; store approvals in user-local file

## TL;DR

The `allowExecutables` default-deny gate (introduced in #1723) previously
enforced only `hooks` and `bin`, leaving MCP server config writes and canvas
deploys outside its control. This PR promotes `mcp` and `canvas` to first-class
enforced types and removes `--trust-canvas-extensions`, unifying all four
executable surfaces under a single approval model. It also fixes a supply-chain
security threat: `apm approve` previously wrote grants into `apm.yml` (committed),
meaning anyone cloning a project inherited all executable trusts automatically.
Approvals now live in `~/.apm/approvals.yml` — user-local, never committed.

> [!IMPORTANT]
> **Breaking changes:**
> `--trust-canvas-extensions` is removed. Consumers of dependency-provided canvases
> must add an `allowExecutables: {}` block to `apm.yml` and run `apm approve <pkg>`.
>
> Existing `allowExecutables` entries in `apm.yml` still work for CI pipelines.
> Personal approvals from interactive `apm approve` sessions now go to `~/.apm/approvals.yml`.

## Problem (WHY)

- [x] **MCP was tracked but not enforced.** `security/executables.py` marked MCP as
  "Reserved for future enforcement" — a self-described gap. A dependency could write
  an MCP config entry without any approval step.
- [x] **`--trust-canvas-extensions` was inconsistent with the gate model.** The flag
  was a one-shot CLI opt-in that bypassed the `allowExecutables` table, so canvas
  trust was invisible to `apm deny` and to any future policy enforcement.
- [!] **Defence-in-depth gap: a dep named `_local`.** `check_executable_approval`
  short-circuited for `package_name == "_local"` before checking `is_first_party`,
  meaning a malicious dependency named `_local` with `is_first_party=False` would
  bypass the canvas gate via the name shortcut alone.
- [!] **`apm approve` wrote to project `apm.yml` — a supply-chain security threat.**
  Committed approvals mean any developer cloning the repo inherits all executable
  trusts automatically. One developer's `apm approve <malicious-pkg>` would silently
  grant execution rights across the entire team.

The approval gate was designed to deliver
["No arbitrary code execution"](https://microsoft.github.io/apm/enterprise/security/)
for all executable surfaces — MCP server processes and canvas Node.js bundles
both qualify and both need the same governance; and approval must remain personal.

## Approach (WHAT)

| # | Change |
|---|--------|
| 1 | Add `EXEC_TYPE_CANVAS`; move `EXEC_TYPE_MCP` from "tracked" to `ENFORCED_EXEC_TYPES` — four types, one gate |
| 2 | `check_executable_approval` returns 4-tuple `(hooks_ok, bin_ok, mcp_ok, canvas_ok)` |
| 3 | Canvas gate re-derives approval via `is_package_approved` independently of `_local` name shortcut |
| 4 | MCP gate added after `collect_transitive` in `commands/install.py` |
| 5 | Remove `--trust-canvas-extensions` / `trust_canvas` from all CLI options, dataclasses, and pipeline kwargs |
| 6 | `apm approve` writes to `~/.apm/approvals.yml` (user-local); `effective_allow_executables()` merges project gate signal + user approvals at install time |

## Implementation (HOW)

- **`security/executables.py`** — `EXEC_TYPE_CANVAS` added; `ENFORCED_EXEC_TYPES` holds all four; `ExecutableDeclaration` gains `canvas_count`/`canvas_details`; scanner detects `.apm/extensions/*/extension.mjs`; new helpers `get_user_approvals_path`, `load_user_approvals`, `save_user_approvals`, `effective_allow_executables`.
- **`install/exec_gate.py`** — `check_executable_approval` expanded to 4-tuple.
- **`install/services.py`** — 4-tuple destructured; canvas gate uses `is_first_party` + explicit `is_package_approved` call (bypasses `_local` name shortcut for defence-in-depth).
- **`install/service.py`** — removed stale `trust_canvas=request.trust_canvas` kwarg.
- **`commands/install.py`** — `--trust-canvas-extensions` and `InstallOptions.trust_canvas` removed; MCP gate inserted after `collect_transitive`; MCP lookup uses `effective_allow_executables()`.
- **`install/context.py`, `request.py`, `pipeline.py`** — `trust_canvas` field and param removed throughout.
- **`commands/approve.py`** — `_load_allow_executables()` reads from `~/.apm/approvals.yml` via `get_user_approvals_path()`; `_approve_all_pending` / `_approve_packages` write via `save_user_approvals()`; `deny_cmd` no longer requires project manifest.
- **`install/phases/integrate.py`** — interactive approval prompt seeds from `load_user_approvals()` and persists via `save_user_approvals()`.
- **`install/template.py`** — `_effective_allow(ctx)` helper merges project gate signal with user approvals.
- **Docs (12 files)** — `approve.md`, `canvas.md`, `security.md`, `primitives-and-targets.md`, `install.md`, `unpack.md`, `experimental.md`, `governance.md`, `package-authoring.md`, `commands.md`, `apm-contributor-dashboard/README.md` updated.
- **Tests (11 files)** — `test_executables.py`, `test_canvas_integrator.py`, `test_install_local_bundle.py`, `test_unpacker.py`, `test_models_validation_rules.py`, `test_install_context.py`, `test_approve_deny.py` updated.

## Diagrams

Gate chain after this PR — new nodes (dashed) show the MCP and canvas gates added to the existing hooks gate.

```mermaid
flowchart LR
    A["scan_package_executables"] --> B["4-tuple result"]:::new
    B --> C{"hooks blocked?"}
    B --> D{"mcp blocked?"}:::new
    B --> E{"canvas blocked?"}:::new
    C -->|yes| F["skip + warn apm approve"]
    C -->|no| G["deploy hooks"]
    D -->|yes| H["skip MCP config write"]:::new
    D -->|no| I["deploy MCP config"]
    E -->|yes| J["skip canvas deploy"]:::new
    E -->|no| K["deploy .github/extensions/"]
    classDef new stroke-dasharray: 5 5;
    class B,D,E,H,J new;
```

Approval storage split — project file enables the gate and holds CI grants; user-local file holds personal approvals.

```mermaid
flowchart LR
    P["apm.yml\n(allowExecutables: {})"]
    U["~/.apm/approvals.yml\n(user-local, not committed)"]
    E["effective_allow_executables()"]
    I["apm install\nexecutable gate"]
    P -->|"gate signal + CI grants"| E
    U -->|"personal approvals"| E
    E -->|"merged approvals"| I
```

## Trade-offs

- **`--trust-canvas-extensions` removed (not deprecated).** Breaking change; callers must migrate to `allowExecutables` + `apm approve`.
- **`apm approve --all` now writes only to `~/.apm/approvals.yml`.** The old CI pattern of `apm approve --all && git add apm.yml` no longer works automatically. CI pipelines must either commit grants directly in `apm.yml` or run `apm approve` in a step that has write access to the runner's home directory. Documented in `approve.md`.
- **`deny_cmd` no longer requires project manifest.** `apm deny` can now be run from any directory. This is intentional — revoking a trust grant should not depend on having the project checked out.
- **`apm unpack` (deprecated) canvas gate simplified.** Canvas deploys freely when the `canvas` feature flag is ON in `apm unpack`, since the deprecated command has no project context for `allowExecutables` lookup. Narrow surface, accepted regression.

## Benefits

1. All four executable surfaces (hooks, bin, mcp, canvas) governed by one `allowExecutables` table — one `apm approve`, one audit trail.
2. Cloning a project with `allowExecutables: {}` no longer auto-grants any executable trust; each developer must explicitly approve packages.
3. Dependency-named-`_local` spoofing of the canvas gate is closed.
4. 17,342 unit tests pass; ruff, pylint R0801, and auth-signals lint all green.

## Validation

<details><summary>Lint chain (all clean)</summary>

```
$ uv run --extra dev ruff check src/ tests/
All checks passed!
$ uv run --extra dev ruff format --check src/ tests/
1295 files already formatted
$ uv run --extra dev python -m pylint --disable=all --enable=R0801 \
    --min-similarity-lines=10 --fail-on=R0801 src/apm_cli/
Your code has been rated at 10.00/10
$ bash scripts/lint-auth-signals.sh
[+] auth-signal lint clean
```

</details>

<details><summary>Test suite (17,342 passed)</summary>

```
$ uv run --extra dev pytest tests/unit/ -q
17342 passed, 2 skipped, 21 xfailed, 19 warnings, 84 subtests passed in 55.64s
```

</details>

### Scenario Evidence

| # | Scenario (user promise) | Principle | Test(s) | Type |
|---|------------------------|-----------|---------|------|
| 1 | No `allowExecutables` block — gate disabled, hooks/bin/mcp/canvas deploy freely (backward compat) | DevX | `test_canvas_deploys_when_allow_executables_none` | unit |
| 2 | Dependency canvas blocked when package not in approvals | Secure by default | `test_dispatch_dependency_named_local_is_not_first_party` | unit |
| 3 | Dependency named `_local` with `is_first_party=False` still blocked by canvas gate | Secure by default | `test_dispatch_dependency_named_local_is_not_first_party` | unit |
| 4 | `apm approve <pkg>` writes grant to `~/.apm/approvals.yml`, NOT to `apm.yml` | Supply-chain safety | `test_approve_all_writes_user_file` | unit |
| 5 | `apm deny <pkg>` removes from user file, not project file | Supply-chain safety | `test_deny_existing_entry` | unit |

## How to test

- [ ] Add `allowExecutables: {}` to a test `apm.yml`; run `apm approve <pkg>`; confirm `~/.apm/approvals.yml` is updated and `apm.yml` is unchanged.
- [ ] Run `apm install --target copilot` and confirm approved canvas deploys to `.github/extensions/`.
- [ ] Remove the approval; re-run; confirm canvas is skipped and diagnostic says `apm approve`.
- [ ] Confirm `apm install --trust-canvas-extensions` exits with "unrecognized option" (flag is gone).
- [ ] Run `uv run --extra dev pytest tests/unit/ -q` — confirm 0 failures.

Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>

### CODE REVIEW

File: src/apm_cli/commands/install.py [hunk 1]
```diff
@@ -1862,7 +1868,35 @@ def _install_apm_packages(ctx, outcome):
...
+            _pkg_slug = getattr(_mcp_dep, "source_package", None) or getattr(
+                _mcp_dep, "package_id", None
+            )
+            if _pkg_slug and not is_package_approved(_allow_execs, _pkg_slug, EXEC_TYPE_MCP):
+                logger.verbose_detail(
+                    f"Skipping MCP server from '{_pkg_slug}': not approved in allowExecutables. "
+                    f"Run 'apm approve {_pkg_slug}' to approve."
+                )
+            else:
+                _filtered_mcp.append(_mcp_dep)
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply src/apm_cli/commands/install.py]: Fixed in 83d0f3bc: `MCPDependency` exposes `name` (not `source_package` or `package_id`), so changed the slug extraction to `_pkg_slug = _mcp_dep.name`. The gate now correctly filters transitive MCP servers whose package name is not approved in `allowExecutables`.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: src/apm_cli/commands/install.py [hunk 2]
```diff
@@ -1266,6 +1259,21 @@ def install(  # noqa: PLR0913
...
+                    from ..security.executables import parse_allow_executables
+                    from ..utils.yaml_io import load_yaml as _load_yaml
+
+                    _apm_yml_path = Path(root or ".") / "apm.yml"
+                    if _apm_yml_path.is_file():
+                        _apm_data = _load_yaml(_apm_yml_path)
+                        if isinstance(_apm_data, dict):
+                            _allow_execs_for_bundle = parse_allow_executables(_apm_data)
+                except Exception:
+                    pass
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply src/apm_cli/commands/install.py]: Fixed in 83d0f3bc: narrowed the bare `except Exception: pass` to log a warning and set `_allow_execs_for_bundle = {}` (empty dict = gate enabled, nothing approved). This is fail-closed: a malformed `apm.yml` no longer silently disables the gate; instead all executables are blocked until the file is fixed.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: src/apm_cli/install/services.py [hunk 3]
```diff
@@ -227,6 +227,24 @@ def _log_hooks_skip(
         )
 
 
+def _log_canvas_skip(package_name: str, package_info: Any, logger: InstallLogger | None) -> None:
+    """Warn about skipped canvas extensions when the package ships them."""
+    _install = Path(package_info.install_path)
+    extensions_root = _install / ".apm" / "extensions"
+    has_canvas = extensions_root.is_dir() and any(
+        (d / "extension.mjs").is_file() for d in extensions_root.iterdir() if d.is_dir()
+    )
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply src/apm_cli/install/services.py]: Fixed in 83d0f3bc: wrapped the `extensions_root.iterdir()` call in `_log_canvas_skip()` with `try/except OSError` — an unreadable `.apm/extensions/` directory now sets `has_canvas = False` (no warning, no abort) rather than propagating and killing the install.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: docs/src/content/docs/integrations/canvas.md [hunk 4]
```diff
@@ -71,35 +71,43 @@ is not picked up mid-session.
 ## Trust gate for dependency canvases
 
 A canvas shipped by a **dependency** is arbitrary executable Node.js code. APM
-blocks dependency-provided canvases by default. To deploy them, opt in
-explicitly:
+blocks dependency-provided canvases by default. To deploy them, add the package
+to the `allowExecutables` block in `apm.yml` and approve it:
+
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply docs/src/content/docs/integrations/canvas.md]: Fixed in 83d0f3bc: updated `canvas.md` to say "When the project opts in to the executable gate (by adding `allowExecutables:` to `apm.yml`), APM blocks dependency-provided canvases unless the package has been explicitly approved." -- removes the misleading "blocked by default" framing and makes the opt-in nature explicit.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: packages/apm-guide/.apm/skills/apm-usage/governance.md [hunk 5]
```diff
@@ -175,26 +175,27 @@ environments where plugin executables are not trusted by default.
...
 extension under `.apm/extensions/<name>/extension.mjs` (executable Node.js).
 Because a canvas from a dependency is arbitrary executable code, APM **blocks
-dependency-provided canvases by default**: the consumer must pass
-`--trust-canvas-extensions` to deploy them. A first-party canvas in the root
-package being installed deploys once the flag is on; dependency canvases always
-require the trust flag.
+dependency-provided canvases by default**: the consumer must declare the package
+in the `allowExecutables` block and run `apm approve <pkg>` to deploy it. A
+first-party canvas in the root package being installed deploys once the flag is
+on; dependency canvases always require explicit approval.
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply packages/apm-guide/.apm/skills/apm-usage/governance.md]: Fixed in 83d0f3bc: updated `governance.md` to read "APM blocks dependency-provided canvases when the project opts in to the executable gate" -- drops the "by default" wording that implied always-on enforcement.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: docs/src/content/docs/enterprise/security.md [hunk 6]
```diff
@@ -30,7 +30,7 @@ APM has no runtime footprint. Once `apm install` or `apm compile` completes, the
...
-- **No arbitrary code execution.** APM does not execute scripts from packages, evaluate expressions in templates, or run downloaded code. (**Canvas exception:** the experimental `canvas` primitive deploys executable `extension.mjs` (Node.js) code to `.github/extensions/` or `~/.copilot/extensions/`; this surface is gated by both the `canvas` experimental flag and `--trust-canvas-extensions` for dependency-provided canvases. See [Canvas extensions](/apm/integrations/canvas/).)
+- **No arbitrary code execution.** APM does not execute scripts from packages, evaluate expressions in templates, or run downloaded code. (**Canvas exception:** the experimental `canvas` primitive deploys executable `extension.mjs` (Node.js) code to `.github/extensions/` or `~/.copilot/extensions/`; this surface is gated by both the `canvas` experimental flag and the `allowExecutables` approval gate for dependency-provided canvases. See [Canvas extensions](/apm/integrations/canvas/).)
 - **No access to application data.** APM never reads databases, API responses, application state, or user data.
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply docs/src/content/docs/enterprise/security.md:34]: Fixed in 83d0f3bc: updated `security.md` bullet to say "when the project opts in via `allowExecutables:` in `apm.yml`, the `allowExecutables` approval gate for dependency-provided canvases" -- accurately describes the opt-in model instead of implying the gate is always active.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

**Comment by danielmeppiel** [comment]
## Advisory review + governance decision

Hi @sergio-sisternes-epam -- this is a consolidated advisory synthesis (apm-review-panel + a 4-expert enterprise-governance panel arbitrated by the CEO persona). It is advisory: nothing here blocks merge. CI is green; the `BLOCKED` state is review-gating only.

### Stance: ship_with_followups

The narrow security wins are strictly positive and should land: default-deny consent expanded to hooks/bin/mcp/canvas, and personal approvals moved off the committed project file. Two small must-fixes before merge, then a larger design follow-up that this PR should **not** be blocked on.

### Must-fix before merge (CI-invisible)

1. **Stale integration test contradicts the new source.** `src/apm_cli/security/executables.py` now sets `ENFORCED_EXEC_TYPES = (HOOKS, BIN, MCP, CANVAS)`, but `tests/integration/test_executables_gate_integration.py` was not updated and still asserts the opposite:
   - line 511: `assert EXEC_TYPE_MCP not in ENFORCED_EXEC_TYPES`
   - line 136: `assert EXEC_TYPE_MCP not in decl.exec_types`

   This is invisible on this PR because `ci.yml` runs only `tests/unit`; the integration suite runs in a separate workflow. It will break the integration pipeline the moment it is next collected. Please update these assertions to match the new enforced set.
2. **Missing CHANGELOG entry.** This is a user-visible hardening (and removes `--trust-canvas-extensions`, which is breaking). Please add a CHANGELOG line under `[Unreleased]`, and call out the flag removal explicitly.

### One design ask for this PR

- **Do not persist a new top-level `~/.apm/approvals.yml`.** Fold personal consent into the **existing** `~/.apm/config` under a namespaced key. This is the single change that keeps net-new control-surface files at zero (see the design issue below for the unified vocabulary).

### Two governance gaps this PR surfaces (follow-up, not blockers)

- **Gap A -- asymmetry.** The policy plane (`apm-policy.yml`) can centrally **DENY** executables (`bin_deploy.deny_all/deny`) but has **no field to GRANT or mandate** them, and `allowExecutables` does not inherit. An admin cannot roll a required hook/MCP out to a large fleet. `effective_allow_executables()` reads only project `apm.yml` + the user file and never consults policy, with **user winning** on conflict.
- **Gap B -- audit false-positive.** A `require`d package whose only payload is executable is gate-stripped to empty `deployed_files`, so the `required-packages-deployed` audit (`policy_checks.py:~198`, `if not locked.deployed_files`) reports a **false** non-compliance. The audit should assert package **presence**, not deployed-executable files on disk.

### Where the bigger design goes

These two gaps, plus a control-surface/vocabulary unification (`allowExecutables` / `executables` / `approvals` / `bin_deploy` collapse onto one noun), the precedence model (deny-wins; org is the ceiling on DENY; consent folds into existing config), and the scaling tooling (`apm explain`), are captured in a dedicated **Executable Trust Governance** design issue being filed now. This PR's narrow wins are independent of that work and can merge once the two must-fixes land.

Thanks for the hardening work here -- the gate expansion is the right direction.

**Review by danielmeppiel — DISMISSED** [review]
Must-fix before merge (CI-invisible)
Stale integration test contradicts the new source. src/apm_cli/security/executables.py now sets ENFORCED_EXEC_TYPES = (HOOKS, BIN, MCP, CANVAS), but tests/integration/test_executables_gate_integration.py was not updated and still asserts the opposite:

line 511: assert EXEC_TYPE_MCP not in ENFORCED_EXEC_TYPES
line 136: assert EXEC_TYPE_MCP not in decl.exec_types
This is invisible on this PR because ci.yml runs only tests/unit; the integration suite runs in a separate workflow. It will break the integration pipeline the moment it is next collected. Please update these assertions to match the new enforced set.

Missing CHANGELOG entry. This is a user-visible hardening (and removes --trust-canvas-extensions, which is breaking). Please add a CHANGELOG line under [Unreleased], and call out the flag removal explicitly.

**Comment by danielmeppiel** [comment]
@danielmeppiel both must-fix change requests are addressed in `e40778fb8`:

**1. Stale integration test (CI-invisible).** `tests/integration/test_executables_gate_integration.py` now matches the new source:
- Enforced-set assertions flipped: `test_mcp_not_in_enforced` -> `test_mcp_and_canvas_in_enforced` (asserts both `EXEC_TYPE_MCP` and `EXEC_TYPE_CANVAS` are in `ENFORCED_EXEC_TYPES`).
- `test_scan_mcp_from_apm_yml` now asserts `has_executables is True` and `EXEC_TYPE_MCP in decl.exec_types`.
- While in there I found the same suite still destructured `check_executable_approval` as a 2-tuple (now a 4-tuple) and routed `apm approve`/`deny` through the project `apm.yml` (now user-local `~/.apm/approvals.yml`) -- 13 collection-time failures total. Folded all of them: 4-tuple destructure, approve/deny via the patched user approvals path, and `deny` no longer asserting a required manifest. Full file is green locally (71 passed).

**2. CHANGELOG.** Added under `[Unreleased]`:
- `### Security` -- gate expansion to mcp/canvas and the user-local approvals store.
- `### Removed` -- **breaking** removal of `--trust-canvas-extensions`, called out explicitly with the migration note.

CI is green on the new head. The PR is `MERGEABLE` but still `BLOCKED` on this review -- re-review / dismiss when you're satisfied.

**Comment by danielmeppiel** [comment]
## APM Review Panel: `ship_with_followups`

> Unifies all four executable surfaces (hooks, bin, mcp, canvas) under one fail-closed allowExecutables gate with user-local approvals -- zero blocking findings, strong convergence on follow-up hardening.

cc @sergio-sisternes-epam @danielmeppiel -- a fresh advisory pass is ready for your review.

Nine panelists reviewed this security PR and returned zero blocking-severity findings. Supply-chain-security confirms the expanded gate is fail-closed for all four exec types, with no default-open or bypass path found (the falsy-name edge in the MCP filter is the closest thing to a hole and is a low-exploitability nit). Auth-expert confirms auth-neutral: no credential, token, or auth-resolver surface is touched. Performance-expert confirms no hot-path regression. The panel converges on four follow-up themes with unusually high independent-rediscovery signal: (1) MCP filter warning omits the `apm approve <pkg>` remediation (raised by cli-logging, devx-ux independently); (2) stale docstrings still referencing apm.yml or the removed trust flag (python-architect, cli-logging, devx-ux, doc-writer all flagged variants); (3) doc-writer found a genuine safety-boundary documentation gap -- unpack.md lost its canvas documentation entirely and now deploys canvas extension.mjs gate-free under the experimental flag with no warning to readers; (4) test-coverage-expert identified three new security-boundary functions (scan_package_executables canvas detection, effective_allow_executables merge, filter_mcp_by_allow_executables) with zero unit tests.

I weight the test-coverage gaps highest. All three functions sit on the `secure_by_default` surface: scan is the entry point that decides whether canvas executables exist, effective_allow_executables is the merge function that determines whether a user-local grant overrides a project deny, and filter_mcp is the hot-path gate for MCP server deployment. Each has missing unit-tier evidence. Per arbitration discipline, missing tests on a secure-by-default surface rank above recommended-tier opinion findings -- a precedence regression in the merge function or a broken glob in the canvas scanner would silently weaken the gate with no automated guardrail catching it. These should be the first follow-up issue filed post-merge. The unpack.md gap is the second-highest signal: it documents a real behavior delta (canvas deploys gate-free on unpack) that the PR silently introduced and that no other doc page warns about. The MCP remediation omission is third -- a user-facing UX defect but not a security boundary issue.

**Dissent.** No direct dissent. Python-architect flagged the 4-tuple's two dead values as cleanup debt while supply-chain-security did not flag it as a bypass risk -- compatible views (dead code is not a vulnerability, but it obscures auditability). A standalone `class`/underscore-prefix comment is a nit, not a blocker.

**Aligned with:** Secure by default (gate is fail-closed for all four exec types; the breaking removal of --trust-canvas-extensions eliminates an opt-out that weakened the default-deny posture). Governed by policy (allowExecutables is the project policy surface; user-local approvals ensure cloning a repo never silently inherits trust). Pragmatic as npm (apm approve / apm deny mirrors the npm audit fix mental model). OSS community-driven (breaking change communicated with a CHANGELOG migration line; CI path preserves automation users).

**Growth signal.** This PR is release-beat gold: "secure-by-default for every executable surface" compounds APM's no-arbitrary-code-execution brand promise and is a concrete differentiator against incumbents that silently run post-install scripts. Lead the next release note with the unified-gate narrative; the user-local approvals angle ("cloning a repo never silently inherits trust") is enterprise-grade positioning. Include a before/after migration snippet in the CHANGELOG breaking entry to make the upgrade feel like a 10-second task.

### Panel summary

| Persona | B | R | N | Takeaway |
|---|---|---|---|---|
| Python Architect | 0 | 2 | 2 | 4-tuple is sound but 2 of 4 values are dead at the call site; canvas re-derivation is correct defence-in-depth. |
| CLI Logging Expert | 0 | 2 | 3 | MCP filter warning omits remediation; deny "not found" message names the stale apm.yml block. |
| DevX UX Expert | 0 | 3 | 2 | Solid unification of four surfaces; MCP warning lacks recovery command; CHANGELOG needs a CI migration line. |
| Supply Chain Security | 0 | 0 | 4 | No default-open or bypass; gate is fail-closed; four hardening nits on the user-local store + MCP falsy-name edge. |
| OSS Growth Hacker | 0 | 1 | 2 | Breaking change well-communicated; unified-gate narrative is a release beat. No funnel risk. |
| Auth Expert | 0 | 0 | 0 | Auth-neutral: no token, credential, or AuthResolver surface touched. |
| Doc Writer | 0 | 2 | 1 | Migration thorough and CHANGELOG correct; two drift gaps (unpack.md canvas, package-authoring trust-flag). |
| Test Coverage Expert | 0 | 3 | 0 | Canvas scan + MCP filter + merge function have no tests; approve/deny and 4-tuple are well-covered. |
| Performance Expert | 0 | 0 | 1 | No hot-path regression; per-dep YAML re-read is noise -- one cache-once nit. |

> B = blocking-severity findings, R = recommended, N = nits.
> Counts are signal strength, not gates. The maintainer ships.

### Top 5 follow-ups

1. **[Test Coverage Expert]** Add unit tests for the three new security-boundary functions: scan_package_executables canvas detection, effective_allow_executables merge precedence, filter_mcp_by_allow_executables -- all sit on the secure-by-default surface with missing evidence; a precedence regression or broken canvas glob would silently weaken the gate with no guardrail.
2. **[Doc Writer]** Restore canvas documentation in unpack.md and add a caution that apm unpack deploys canvas extension.mjs with no allowExecutables gate under the experimental flag -- a reader currently gets no warning that an unpacked bundle's executable canvas deploys gate-free.
3. **[CLI Logging Expert]** Add the `apm approve <pkg>` remediation to the MCP filter summary warning, matching hooks/canvas skip messages -- convergent finding; every action-blocking warning should name the recovery command.
4. **[DevX UX Expert]** Add a CI migration path to the CHANGELOG breaking entry (before/after snippet for --trust-canvas-extensions removal) -- CI has no interactive apm approve; name the apm.yml pre-approval alternative.
5. **[Supply Chain Security Expert]** Invert the falsy-name guard in filter_mcp_by_allow_executables so a dep with None/empty slug is treated as not-approved -- low exploitability but the gate should not have a structural hole on a falsy edge; one-line fix.

### Architecture

```mermaid
classDiagram
    direction LR
    class ExecutableDeclaration {
        +package_key str
        +hook_count int
        +mcp_count int
        +bin_count int
        +canvas_count int
        +has_executables bool
        +exec_types list
    }
    class check_executable_approval {
        +__call__(pkg, info, allow, ctx) tuple4
    }
    class is_package_approved {
        +__call__(allow, key, type) bool
    }
    class effective_allow_executables {
        +__call__(project) dict_or_None
    }
    class load_user_approvals {
        +__call__() dict
    }
    class save_user_approvals {
        +__call__(approvals) None
    }
    class filter_mcp_by_allow_executables {
        +__call__(deps, allow, logger) list
    }
    class integrate_package_primitives
    check_executable_approval ..> is_package_approved : per type
    effective_allow_executables ..> load_user_approvals : merges user file
    filter_mcp_by_allow_executables ..> effective_allow_executables : merges
    filter_mcp_by_allow_executables ..> is_package_approved : checks MCP
    integrate_package_primitives ..> check_executable_approval : 4-tuple gate
    integrate_package_primitives ..> is_package_approved : canvas re-derive
    class check_executable_approval:::touched
    class effective_allow_executables:::touched
    class load_user_approvals:::touched
    class save_user_approvals:::touched
    class filter_mcp_by_allow_executables:::touched
    class ExecutableDeclaration:::touched
    class integrate_package_primitives:::touched
    classDef touched fill:#fff3b0,stroke:#d47600
```

```mermaid
flowchart TD
    A["apm install"] --> D["read apm.yml allowExecutables + ~/.apm/approvals.yml"]
    D --> E{"allowExecutables block present?"}
    E -- "None: gate disabled" --> F["All executables deploy (backward compat)"]
    E -- "dict: gate enabled" --> G["effective_allow_executables() merge project + user-local"]
    G --> I["integrate_package_primitives()"]
    I --> J["check_executable_approval() -> hooks,bin,mcp,canvas 4-tuple"]
    J --> K{"hooks approved?"}
    K -- No --> L["skip hook deploy + warn"]
    K -- Yes --> M["HookIntegrator"]
    J --> N{"canvas + not first-party?"}
    N -- Yes --> O{"is_package_approved re-derive canvas gate"}
    O -- No --> P["skip canvas deploy + warn"]
    O -- Yes --> Q["CanvasIntegrator"]
    A --> V["filter_mcp_by_allow_executables()"]
    V --> W{"MCP approved?"}
    W -- No --> X["MCP server filtered out"]
    W -- Yes --> Y["MCPIntegrator"]
    A2["apm approve pkg"] --> AA["save_user_approvals() -> ~/.apm/approvals.yml mode 0600"]
```

### Recommendation

Merge this PR. Zero blocking findings across nine panelists; the expanded gate is fail-closed and supply-chain-verified. File a single follow-up issue post-merge covering the three missing security-boundary unit tests (highest signal) and the unpack.md documentation gap (second-highest). The MCP remediation warning, CHANGELOG CI migration snippet, and falsy-name guard inversion are low-effort fixes that can land in the same follow-up PR or independently. The maintainer should clear the CHANGES_REQUESTED review and merge at will.

---

<sub>This panel is advisory. It does not block merge. Re-apply the `panel-review` label after addressing feedback to re-run.</sub>

**Comment by danielmeppiel** [comment]
## apm-review-panel advisory (terminal pass) -- PR #1865

Final converged pass on the `allowExecutables` gate expansion (MCP +
canvas; user-local approvals). Nine specialist personas plus the CEO
synthesizer. **CEO stance: `ship_now`** -- zero blocking findings; every
`recommended`-tier item folded into this PR.

### Folded in this run

- (python-architect / supply-chain-security) `load_user_approvals` shape
  validation now rejects non-`str` inner keys -- fail-closed against a
  hand-tampered `approvals.yml` -- resolved in `a5949ebc3`.
- (supply-chain-security / test-coverage) malformed-entry shape-drop test
  added, mutation-break proven -- resolved in `a5949ebc3`.
- (test-coverage) `~/.apm` directory-mode `0o700` test added -- resolved in
  `a5949ebc3`.
- (devx-ux / cli-logging) unnamed-MCP-dep remediation no longer prints a
  non-runnable `apm approve <unnamed>`; the branch now tells the user to
  identify the dep in `apm.yml` -- resolved in `a5949ebc3`.
- (cli-logging / devx-ux) integrate-phase persisted-count message clarified
  to "(N newly approved)" -- resolved in `a5949ebc3`.
- (performance) `import contextlib` hoisted to module scope in
  `template.py` -- resolved in `a5949ebc3`.
- (oss-growth) CHANGELOG entry reframed to lead with the security upgrade
  ("Breaking (security)") -- resolved in `a5949ebc3`.
- (oss-growth) `unpack.md` cross-links the Executable approval reference --
  resolved in `a5949ebc3`.

### Deferred (out-of-scope follow-ups)

- (python-architect) declare `_effective_allow_cache` as a real dataclass
  field -- scope boundary: cosmetic; the `setattr` on a non-frozen
  `InstallContext` is functionally correct, so this is zero-behavior polish.
- (devx-ux) mention the gate's exit code in the CHANGELOG -- scope boundary:
  belongs in the `approve.md` reference, not the changelog lead.
- (test-coverage) assert the integrate count-message wording in a test --
  scope boundary: nit; constructing the ctx + blocked-executables + prompt
  mock is disproportionate to a wording check already exercised indirectly.

### Regression-trap evidence (mutation-break gate)

- `TestLoadUserApprovalsShapeValidation::test_drops_malformed_entries` --
  deleted the inner-key `str` guard (`and all(isinstance(k, str) for k in
  entry)`); test FAILED as expected; guard restored.

### Lint contract

`uv run --extra dev ruff check src/ tests/` and `ruff format --check
src/ tests/` both silent; pylint R0801 10.00/10; auth-signal lint clean.

### CI

All required checks green at `a5949ebc3` (Lint, Build & Test Shard 1/2,
gate, build, PR Binary Smoke, Analyze python/actions, CodeQL, Spec
conformance, Coverage Combine; `deploy` skipped by design). 0 CI fix
iterations. 2799 unit tests pass locally.

### Mergeability status

| PR | head SHA | CEO stance | iters | folds | defers | CI | mergeable | mergeStateStatus |
|----|----------|------------|-------|-------|--------|----|-----------|------------------|
| #1865 | `a5949ebc` | ship_now | 2 | 8 | 3 | green | MERGEABLE | BLOCKED |

`BLOCKED` reflects the outstanding required-review gate (the prior
CHANGES_REQUESTED review is now dismissed); the advisory panel does not
gate merge. Ready for maintainer review.

### Convergence

2 outer panel iterations; 0 Copilot rounds with open inline comments.
Final panel verdict: `ship_now`.
